### ISSUE TITLE
`ImplOwner` method's receiver changes from `&self` to `&mut self`

### ISSUE THREAD

**Issue by obi1kenobi** [obi1kenobi/cargo-semver-checks#1223]

**Comment by KhaledSayed04** [comment]
For this issue, what should happen when there is another method with the same name still takes immutable reference? the lint shouldn't report this as a breaking change? or it should detect this by the order & type & number of parameters?

Old code:
`impl Foo {
    fn foo(&self) {}
    fn foo(&self, x: i32) {}
}`

new code:
`impl Foo {
    fn foo(&mut self) {} 
    fn foo(&self, x: i32) {}
}`

**Comment by obi1kenobi** [comment]
I believe the proposed new code wouldn't compile: that looks to me like a
duplicate definition compile error in today's Rust.

Good question though! Thinking through edge cases like this is exactly the
energy we need in this project.

**Comment by KhaledSayed04** [comment]
Is this a valid id for this issue? "method_receiver_ref_mutability_increased" or "method_receiver_self_ref_mutability_increased" to differentiate it from the case where the kind is Box<Self> or any other kind?

**Comment by KhaledSayed04** [comment]
I opened a pull request for this issue. I would appreciate it if you could take a look and give me your feedback.

**Comment by obi1kenobi** [comment]
> Is this a valid id for this issue?

Yeah, this is perfect :+1:

> "method_receiver_ref_mutability_increased" or "method_receiver_self_ref_mutability_increased" to differentiate it from the case where the kind is Box or any other kind?

I find "mutability increased" confusing, because it isn't a continuous spectrum. It's either mutable or not. Naming is hard :)

Another question is whether we want to differentiate between kind `Self` and other ones like `Box` or `Rc`. I think the current way you've written the lint, it will only look for kind `Self`, so changing `self: &Rc<Self>` to `self: &mut Rc<Self>` won't be caught here even though it's also a `&` to `&mut` change. Is this on purpose?

I'd recommend trying to come up with a plan for which cases end up in which lints, then suggest lint names for all cases in this lint group so we can make sure they are all consistent with each other and with other existing lints. I suspect that handling `&Rc<Self>` to `&mut Rc<Self>` will end up being best in this lint and not elsewhere, but I'm not 100% sure and I'm willing to entertain other ideas backed up by good reasons :)

**Comment by KhaledSayed04** [comment]
>I find "mutability increased" confusing, because it isn't a continuous spectrum. It's either
>mutable or not. Naming is hard   :)

so, what about "method_receiver_ref_became_mutable_ref"?

>Is this on purpose?
I thought this is the required at this lint, I will update it to check only they have the same kind instead.

>I'd recommend trying to come up with a plan for which cases end up in which lints, 
>then suggest lint names for all cases in this lint group so we can make sure they are 
>all consistent with each other and with other existing lints.

 
[ImplOwner method's receiver changes from &self to &mut self #1223](https://github.com/obi1kenobi/cargo-semver-checks/issues/1223) 
method_receiver_ref_became_mutable_ref
or
method_receiver_ref_now_mutable_ref

When the receiver ref becomes mutable without changes in kind.  


[ImplOwner method's receiver changes from &self to self #1224]
(https://github.com/obi1kenobi/cargo-semver-checks/issues/1224) 
[ImplOwner method's receiver changes from &mut self to self #1225]
(https://github.com/obi1kenobi/cargo-semver-checks/issues/1225)
I don't know why they need seperate lints, why not only one "method_receiver_by_ref_now_by_value" (this may be confusing because of the by_ref & by_mut_ref at Receiver type) or "method_receiver_change_ref_to_value" ? 

When the method owns self instead of referencing without changes in kind.


[ImplOwner method's receiver changes kind (e.g. from Self to Box<Self> or Pin<&mut Self>; see schema for more examples) #1226]
(https://github.com/obi1kenobi/cargo-semver-checks/issues/1226)
method_receiver_kind_changed 

When the receiver kind becomes different

[ImplOwner method becomes an associated function #1227](https://github.com/obi1kenobi/cargo-semver-checks/issues/1227)
method_became_associated_function
or
method_now_associated_function

When 'self' is omitted in the new version. (According to the schema: method without receiver)

**Comment by KhaledSayed04** [comment]
>won't be caught here even though it's also a & to &mut change. Is this on purpose?

I thought this won't be caught at this lint. There isn't a logical reason to avoid this, so I'll update it to check they have the same kind instead.

----------------------------------------


### [ImplOwner method's receiver changes from &self to &mut self #1223]
method_receiver_ref_became_mut_ref
method_receiver_ref_became_mut
method_receiver_ref_became_mutable
method_receiver_ref_now_mutable
When the method receiver ref becomes mutable. Without changes in kind.


----------------------------------------

### [ImplOwner method's receiver changes from &self to self #1224]
### [ImplOwner method's receiver changes from &mut self to self #1225]

I don't know why not to solve them with a single lint. 
"method_receiver_ref_to_value"
"method_receiver_by_ref_to_by_value" (this may be confusing because the Receiver type contains both by_ref and by_mut_ref)
"method_receiver_now_takes_ownership"
"method_receiver_now_by_value".
When the method takes the ownership of self instead of only borrowing. Without changes in kind.


----------------------------------------


### [ImplOwner method's receiver changes kind (e.g. from Self to Box<Self> or Pin<&mut Self>; see schema for more examples) #1226]

method_receiver_kind_changed


----------------------------------------

### [ImplOwner method becomes an associated function #1227]

method_became_associated_function
method_now_associated_function

**Comment by Frank-III** [comment]
>[ImplOwner method's receiver changes from &self to &mut self https://github.com/obi1kenobi/cargo-semver-checks/issues/1223]
method_receiver_ref_became_mut_ref
method_receiver_ref_became_mut
method_receiver_ref_became_mutable
method_receiver_ref_now_mutable
When the method receiver ref becomes mutable. Without changes in kind.

 we use type  to refer to `ImplOwner` like `type_requires_more_const_generic_params` or `type_marked_deprecated`, should we make them: 
```
type_method_receiver_ref_became_mut_ref
type_method_receiver_ref_became_mut
type_method_receiver_ref_became_mutable
type_method_receiver_ref_now_mutable
```

**Comment by obi1kenobi** [comment]
Resolved by #1492.
